synapse/handlers/sync.py

```diff
@@ -683,12 +683,11 @@ async def get_state_at(
             stream_position: point at which to get state
             state_filter: The state filter used to fetch state from the database.
         """
-        # FIXME this claims to get the state at a stream position, but
-        # get_recent_events_for_room operates by topo ordering. This therefore
-        # does not reliably give you the state at the given stream position.
-        # (https://github.com/matrix-org/synapse/issues/3305)
         last_events, _ = await self.store.get_recent_events_for_room(
-            room_id, end_token=stream_position.room_key, limit=1
+            room_id,
+            end_token=stream_position.room_key,
+            limit=1,
+            order_by="stream",
         )
```

Comment:
We should add a test for https://github.com/matrix-org/synapse/issues/3305 so it doesn't regress.

I don't have clear reproduction steps for this one and maybe we can just consider the test we write for https://github.com/matrix-org/synapse/issues/12281 as a subset of this problem and only need one test 🤷